bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Display.java

```diff
@@ -929,6 +931,16 @@ public void asyncExec (Runnable runnable) {
 	}
 }
 
+@Override
+public void execute(Runnable command) {
+	Objects.requireNonNull(command);
+	if (thread == Thread.currentThread()) {
+		command.run();
+	} else {
```

Comment:
What's the usecase to either run synchronous or asynchronous? You could also offer 3 Method. getSynchronousExecutor, getAsynchronousExecutor, "getRunSomewhenExecutor" ;-) - we could find a better name when we would know the usecase - if any.

Reply:
The idea is that one don't need to know, I always *never* care, e.g. if I want to update the text of a label in what Thread I'm currently running, I just want to update the damn label and as it even do not return a value I even don't care if it happens 'now' and I'm blocked for that while or it happen some microseconds later in another thread.

so this simply prevents me from the usual bloat:

```
if (Display.getCurrent() == null) {
do tomething with the ui
} else {
 Display.getDefault().asyncExec(()->{
  do the same as in the other branch

});
}
```

Reply:
if it does not matter when it runs, you could just always run it async.
The usual problem then is that the code does not know if the widget is already is disposed, one needs to check that first... horrible boiler code. Would be cool to have such a method on a widget, and the code would only run if the widget is not disposed yet ... that would be an improvement!

Reply:
> if it does not matter when it runs, you could just always run it async.

This is simply not true, calling `async` is prone to deadloack if the code tries to wait for the "async" operation to complete, my inital implementation always used async and you can easily deadlock with that aproach. Beside that its a perfomance penalty involved scheduling a runnable async.

> The usual problem then is that the code does not know if the widget is already is disposed, one needs to check that first... horrible boiler code. Would be cool to have such a method on a widget, and the code would only run if the widget is not disposed yet ... that would be an improvement!

Sure, but the goal here is not to improve widget disposal but integrate better with the concurrency framework and there it is important that this distinction is made!

Reply:
Why would code wait for async? then it should take syncexec. the perfomance penalty is neglegible when you want to update a text - compared to drawing the text.

You could add that concurrency framework thing to widget - and would solve two problems with a single call.
Please, convince me that your change is worth by showing a benefit in a nontrivial example  with correct error handling against using the current API in the best manner. You could also use such example for the noteworthy new.
It should be like ... "till now you used this and now you can use that and it's x commands less" - win.

Reply:
> You could add that concurrency framework thing to widget - and would solve two problems with a single call.

This is not about calling a single method, I want to execute code in the UI thread and that's what executors are for, that the caller does not has to fiddle around what thread it is in, what display to acquire and so on.

> You could add that concurrency framework thing to widget

Having such API does not make sense at the widget level as it is not about calling a single widget method.

> Please, convince me that your change is worth by showing a benefit in a nontrivial example with correct error handling

Just because it is not useful for /you/ does not mean it is not useful for others. I have already explained why manual coding does not satisfies my needs and added an explanation to https://github.com/eclipse-platform/eclipse.platform.swt/issues/64 why this is useful, my goal is not to convince specifically **you** to use that APi feel free to ignore it if you think the good old way is the best of all choices ;-)

> "till now you used this and now you can use that and it's x commands less" - win.

Till now I need to use (+ any not show error/dispose handling not show):

```
if (Display.getCurrent() == null) {
do something with the ui
} else {
 Display.getDefault().asyncExec(()->{
  do the same as in the other branch

});
}
```
and now I can use (+ any not show error/dispose handling not show):
```
Display.executor(()->{
do something with the ui
});
```

I even can take a CompletableFuture and in input parameter in

```
void handleError(CompletableFuture<?> future) {
      future.handleAsync((result, error)->{
		if (error != null) {
			//show my favorite error dialog
			
		}
		return result;
	}, Display.executor());
}
```

without having to know what thread the caller uses, when in an unforeseeable future an error might occurs.

All these are "wins" for me and for sure I can just write another `SwtUtil` class (why org.eclipse.swt.widgets.Display.syncCall(SwtCallable<T, E>) was not put into such a static? It does not uses any internal state by the way?) and let other people do the same or just get mad about SWT being stuck at java 1.4 ...

Reply:
These unrealistic trivial example does not convince me. Neither that the executor is helpful nor that it needs to be in Display. 
As an alternative a label.executor(()->label.setText("new")) would be neat to avoid dispose checks.

Reply:
> As an alternative a label.executor(()->label.setText("new")) would be neat to avoid dispose checks.

I don't want to cover a single call, please open another ticket/enahncement request if that's important for you.
I want to execute an arbitrary code block in the ui thread...

From my POV your example is unrealistic and trivial as I rather would like call:
`label.setText("new"):`
and don't care:

1. if I'm on swt thread or not (if not spin an async call for me)
2. label is disposed (as it is not visible I dont care if my call would simply be a noop)

but thats completely out of scope of this cahnge and it is **not** what I'd like to archive. I'D like to archie modern programming primitives instead of all the UI frezze stuff that regular happen because people don't do all the checks.

Here is a simple example that shows that it **is** important to know the thread and why you cant simply always use async exec, and yes I know that in that example I could have know that it might block, but as noted above: I don't want to check all the time if I'm in or out the Ui thread ...

```
public class SnippetBlocking {

	public static void main(String[] args) throws InterruptedException, ExecutionException {
		Display display = new Display();
		Shell shell = new Shell(display);
		shell.setLayout(new FillLayout());
		shell.setText("Snippet 1");
		Button button = new Button(shell, SWT.PUSH);
		button.setText("Press me");
		Button check = new Button(shell, SWT.CHECK);
		check.setText("Do you like to block?");
		button.addSelectionListener(new SelectionListener() {

			@Override
			public void widgetSelected(SelectionEvent e) {
				Executor willBlock = display::asyncExec;
				Executor willNotBlock = r->{
					if (Display.getCurrent()==null) {
						Display.getDefault().asyncExec(r);
					} else {
						r.run();
					}

				};
				Executor ex;
				if (check.getSelection()) {
					ex = willBlock;
				} else {
					ex = willNotBlock;
				}
				CompletableFuture.completedFuture("Hello").thenApplyAsync(str -> str + " Stage 1", ex)
						.thenAccept(System.out::println).join();
			}

			@Override
			public void widgetDefaultSelected(SelectionEvent e) {

			}
		});
		shell.open();
		while (!shell.isDisposed()) {
			System.out.println("Dispatch...");
			if (!display.readAndDispatch())
				display.sleep();
		}
		display.dispose();
	}
}
```

Reply:
> calling async is prone to deadloack if the code tries to wait for the "async" operation to complete

Couldn't resist: sorry, it's join() that is prone to deadlock, not asyncExec(). 

I reviewed lot of code/bugs in UI area in last few years and **never** saw a deadlock caused by asyncExec(). join() and syncExec() are best friends for deadlocks, not asyncExec().

And after deadlocks via join() or syncExec() most prominent error is to **not** check if the widget is already disposed inside the asyncExec() block.

So from this point of view, providing an **async** API that executes a task on a widget **only** it it is not disposed yet would prevent lot of bugs and boiler plated code.

Reply:
> Couldn't resist: sorry, it's join() that is prone to deadlock, not asyncExec().

This is the problem here, but it was claimed that I can always use `asyncExec()` instead of my executor and that is simply not true, I have to know if I'm on the Ui thread or not, and that's what the executor does I just don't need to know. and as described in the issue description this is specifically to integrate with the concurrent API, and the executor is the most basic building block to archive that.

> I reviewed lot of code/bugs in UI area in last few years and **never** saw a deadlock caused by asyncExec().

But you also review a lot of code that deadlocks in certain circumstance and flapping UI test, and thats because we always try to be smart and "know" what happens when, but then there are cases where it is not obvious, we need to wait and alike. That's what one could avoid using the concurrent API but currently SWT requires me to do repetive check if I really like to use it or know the full context of a call chain.

> join() and syncExec() are best friends for deadlocks

True and that's why I'd like to add some easy to use integration with the concurrent API and Completable future, because that allows people to fulfill their goals without the need of ever calling those `syncExec()` and repeatably check for what thread they are in. To make things wore we now not only have `syncExec()` but also `syncCall()` ... still I can understand that people sometimes need to get the result of the operation and there we come to the end beginning again.

> So from this point of view, providing an **async** API that executes a task on a widget **only** it it is not disposed yet would prevent lot of bugs and boiler plated code.

The good think is, that once you have such basic building block you can easily write convenient API on top of it, given that you have a Widget class (I used the Text Widget here) you can now add a method safeCall:

```
public void safeRun(Consumer<? super Text> consumer) {
	getDisplay().execute(() ->{
		if (isDisposed()) {
			return;
		}
		consumer.accept(this);
	});
}
```

and call it like

```
Text text = ...
text.safeRun(t->t.setText("Will never be disposed"));
```

This has now two properties:

1. I can call it fro **any** thread, so no need to check or "know for sure" that you are on an UI thread, this is extremely useful if you implement API / Services where you simply don't know who calls then when as the APi contract simply don't know anything about SWT!
2. If the widget is disposed in the meanwhile, the code will be just a no-op

If you better like a static method use

Reply:
> I do like the idea, but i do not get the usecase where it is better then asyncExec.

@jukzi  Are there any concerns given the examples, possibles use cases and api references quite explain for what cases it might be required to have such a specialized executor?

Reply:
I still think we still need a *good* example. Preferably from within eclipse code where it can be useful. The only situation i know,  where the "compute now or later - i do dont care "- pattern is used is org.eclipse.jface.util.Throttler.throttledExec(). But even there i don't see that a Executor would simplify anything. I still like the idea of supporting the Executor style - but would rather see it in a platform independent class, as there is not need to extend Display. So if you would give an example from within eclipse where it can be used and change the Implementation i would support this kind of API. Another technical detail might be that it should be tested for multi-display usage which we got aware of could be used in win. If on the other hand multi display usage requires a (non static!) method in Display then i would support even a extension of Display.